File: frontend/src/types/property.ts
export interface MarketPrice {
  source: string
  price: number
  price_per_m2: number | null
  deal_date: string | null
}

export interface AnalysisResult {
  market_price: number | null
  gap_amount: number | null
  gap_pct: number | null
  acquisition_tax: number | null
  risk_keywords: string[]
  is_blind_land: boolean
  needs_farm_cert: boolean
  is_safe: boolean
  score: number | null
  tenant_deposit?: number
  analyzed_at: string | null
}

export interface Property {
  id: number
  notice_no: string
  address: string
  sido: string | null
  sigungu: string | null
  property_type: string
  land_category: string | null
  area_m2: number | null
  appraisal_value: number | null
  min_bid_price: number
  fail_count: number
  bid_start_dt: string | null
  bid_end_dt: string | null
  description: string | null
  notice_url: string | null
  image_url: string | null
  is_active: boolean
  created_at: string
  updated_at: string
  analysis: AnalysisResult | null
  market_prices: MarketPrice[]
  is_favorite?: boolean
  is_read?: boolean
}

export interface PropertyListResponse {
  total: number
  page: number
  page_size: number
  items: Property[]
}

export interface SummaryResponse {
  total_properties: number
  safe_properties: number
  avg_gap_pct: number
  top_gap_pct: number
  avg_score: number
  last_synced_at: string | null
  type_distribution: Record<string, number>
  sido_distribution: Record<string, number>
}

export interface TopResponse {
  count: number
  items: Property[]
}

export interface GapDistribution {
  label: string
  count: number
}

export interface FilterState {
  sido: string
  property_type: string
  min_price: string
  max_price: string
  safe_only: boolean
  min_gap_pct: string
  min_fail_count: string
}

File: backend/app/schemas/property.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List
import json


class MarketPriceSchema(BaseModel):
    source: str
    price: int
    price_per_m2: Optional[int] = None
    deal_date: Optional[str] = None

    class Config:
        from_attributes = True


class AnalysisResultSchema(BaseModel):
    market_price: Optional[int] = None
    gap_amount: Optional[int] = None
    gap_pct: Optional[float] = None
    acquisition_tax: Optional[int] = None
    risk_keywords: List[str] = []
    is_blind_land: bool = False
    needs_farm_cert: bool = False
    is_safe: bool = True
    score: Optional[float] = None
    tenant_deposit: Optional[int] = None
    analyzed_at: Optional[datetime] = None

    @classmethod
    def from_orm_with_json(cls, obj):
        data = {
            "market_price": obj.market_price,
            "gap_amount": obj.gap_amount,
            "gap_pct": obj.gap_pct,
            "acquisition_tax": obj.acquisition_tax,
            "risk_keywords": json.loads(obj.risk_keywords) if obj.risk_keywords else [],
            "is_blind_land": obj.is_blind_land,
            "needs_farm_cert": obj.needs_farm_cert,
            "is_safe": obj.is_safe,
            "score": obj.score,
            "tenant_deposit": obj.tenant_deposit,
            "analyzed_at": obj.analyzed_at,
        }
        return cls(**data)

    class Config:
        from_attributes = True


class PropertyBase(BaseModel):
    notice_no: str
    address: str
    sido: Optional[str] = None
    sigungu: Optional[str] = None
    property_type: str
    land_category: Optional[str] = None
    area_m2: Optional[float] = None
    appraisal_value: Optional[int] = None
    min_bid_price: int
    fail_count: int = 0
    bid_start_dt: Optional[datetime] = None
    bid_end_dt: Optional[datetime] = None
    description: Optional[str] = None
    notice_url: Optional[str] = None
    image_url: Optional[str] = None
    is_active: bool = True


class PropertyCreate(PropertyBase):
    pass


class PropertyResponse(PropertyBase):
    id: int
    created_at: datetime
    updated_at: datetime
    analysis: Optional[AnalysisResultSchema] = None
    market_prices: List[MarketPriceSchema] = []
    is_favorite: Optional[bool] = False
    is_read: Optional[bool] = False

    class Config:
        from_attributes = True


class PropertyListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[PropertyResponse]


class SummaryResponse(BaseModel):
    total_properties: int
    safe_properties: int
    avg_gap_pct: float
    top_gap_pct: float
    last_synced_at: Optional[datetime] = None


# 필터 쿼리 파라미터용
class PropertyFilter(BaseModel):
    sido: Optional[str] = None
    property_type: Optional[str] = None
    min_price: Optional[int] = None
    max_price: Optional[int] = None
    safe_only: bool = False
    min_gap_pct: Optional[float] = None
    min_fail_count: Optional[int] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
